src/sysrepo.c

```diff
@@ -6083,38 +6084,64 @@ sr_rpc_internal_input_update(sr_conn_ctx_t *conn, struct lyd_node *input_op)
 {
     sr_error_info_t *err_info = NULL;
     const struct lys_module *ly_mod, *ly_srfd_mod;
+    struct ly_set *set = NULL;
     uint32_t i = 0;
+    struct lyd_node *node;
 
     assert(!strcmp(LYD_NAME(input_op), "factory-reset"));
 
-    if (lyd_child(input_op)) {
-        /* explicit modules specified */
-        return NULL;
-    }
-
     /* find sysrepo-factory-default module */
     ly_srfd_mod = ly_ctx_get_module_implemented(conn->ly_ctx, "sysrepo-factory-default");
     assert(ly_srfd_mod);
 
-    while ((ly_mod = ly_ctx_get_module_iter(conn->ly_ctx, &i))) {
-        if (!ly_mod->implemented) {
-            continue;
-        } else if (!strcmp(ly_mod->name, "sysrepo")) {
-            /* sysrepo internal data will not be reset */
-            continue;
-        } else if (!strcmp(ly_mod->name, "ietf-netconf")) {
-            /* ietf-netconf defines data but only internal that should be ignored */
-            continue;
-        } else if (!sr_module_has_data(ly_mod, 0)) {
-            /* no configuration data */
-            continue;
+    /* check for explicitly defined modules */
+    if ((err_info = sr_lyd_find_xpath(input_op, "modules/module", &set))) {
+        goto cleanup;
+    }
+    if (set->size == 0) {
+        if ((err_info = sr_lyd_new_inner(input_op, ly_srfd_mod, "modules", &node))) {
+            goto cleanup;
         }
 
-        if ((err_info = sr_lyd_new_term(input_op, ly_srfd_mod, "module", ly_mod->name))) {
-            return err_info;
+        while ((ly_mod = ly_ctx_get_module_iter(conn->ly_ctx, &i))) {
+            if (!ly_mod->implemented) {
+                continue;
+            } else if (!strcmp(ly_mod->name, "sysrepo")) {
+                /* sysrepo internal data will not be reset */
+                continue;
+            } else if (!strcmp(ly_mod->name, "ietf-netconf")) {
+                /* ietf-netconf defines data but only internal that should be ignored */
+                continue;
+            } else if (!sr_module_has_data(ly_mod, 0)) {
+                /* no configuration data */
+                continue;
+            }
+
+            if ((err_info = sr_lyd_new_term(node, ly_srfd_mod, "module", ly_mod->name))) {
+                goto cleanup;
+            }
         }
     }
+    ly_set_free(set, NULL);
 
+    /* check for explicitly defined datastores */
+    if ((err_info = sr_lyd_find_xpath(input_op, "datastores/datastore", &set))) {
+        goto cleanup;
+    }
+    if (set->size == 0) {
+        if ((err_info = sr_lyd_new_inner(input_op, ly_srfd_mod, "datastores", &node))) {
+            goto cleanup;
+        }
+        if ((err_info = sr_lyd_new_term(node, ly_srfd_mod, "datastore", "ietf-datastores:startup"))) {
+            goto cleanup;
+        }
+        if ((err_info = sr_lyd_new_term(node, ly_srfd_mod, "datastore", "ietf-datastores:running"))) {
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    ly_set_free(set, NULL);
```

Comment:
Is this really necessary? The modules always needed to be passed to the callback but the datastores should not, unless I am mistaken.

Reply:
nope, you are right